File: ios/Leanplum.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_REMAP_MODULE(Leanplum, RNLeanplum, RCTEventEmitter)
RCT_EXTERN_METHOD(setAppVersion: (NSString *)appVersion)
RCT_EXTERN_METHOD(setAppIdForDevelopmentMode: (NSString *)appId accessKey:(NSString *)accessKey)
RCT_EXTERN_METHOD(setAppIdForProductionMode: (NSString *)appId accessKey:(NSString *)accessKey)
RCT_EXTERN_METHOD(setApiConnectionSettings: (NSString *)hostName servletName:(NSString *)servletName ssl:(BOOL)ssl)
RCT_EXTERN_METHOD(setSocketConnectionSettings: (NSString *)hostName port:(NSInteger *)port)
RCT_EXTERN_METHOD(setDeviceId: (NSString *)id)
RCT_EXTERN_METHOD(setUserId: (NSString *)id)
RCT_EXTERN_METHOD(setUserAttributes: (NSDictionary *)attributes)
RCT_EXTERN_METHOD(userId: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(deviceId: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(start)
RCT_EXTERN_METHOD(hasStarted: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(track: (NSString *)event params:(NSDictionary *)params)
RCT_EXTERN_METHOD(trackPurchase: (NSString *)purchaseEvent value:(double *)value currencyCode:(NSString *)currencyCode purchaseParams:(NSDictionary *)purchaseParams)
RCT_EXTERN_METHOD(trackInAppPurchases)
RCT_EXTERN_METHOD(disableLocationCollection)
RCT_EXTERN_METHOD(setDeviceLocation: (double *)latitude longitude:(double *)longitude type:(NSInteger *)type)
RCT_EXTERN_METHOD(forceContentUpdate)
RCT_EXTERN_METHOD(setVariables: (NSDictionary *)variables)
RCT_EXTERN_METHOD(getVariable: (NSString *)variableName resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getVariables: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(onStartResponse: (RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(onValueChanged: (NSString *)variableName)
RCT_EXTERN_METHOD(onVariablesChanged: (NSString *)listener)
RCT_EXTERN_METHOD(setVariableAsset: (NSString *)name filename:(NSString *)filename)
RCT_EXTERN_METHOD(getVariableAsset: (NSString *)name resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(pauseState)
RCT_EXTERN_METHOD(resumeState)
RCT_EXTERN_METHOD(trackAllAppScreens)
RCT_EXTERN_METHOD(advanceTo: (NSString *)state)
RCT_EXTERN_METHOD(advanceToWithInfo: (NSString *)state info:(NSString *)info)
RCT_EXTERN_METHOD(advanceToWithParams: (NSString *)state params:(NSDictionary *)params)
RCT_EXTERN_METHOD(advanceToWithInfoAndParams: (NSString *)state info:(NSString *)info params:(NSDictionary *)params)
RCT_EXTERN_METHOD(onVariablesChangedAndNoDownloadsPending: (NSString *)listener)
RCT_EXTERN_METHOD(onceVariablesChangedAndNoDownloadsPending: (NSString *)listener)
RCT_EXTERN_METHOD(onMessageDisplayed: (NSString *)listener)
RCT_EXTERN_METHOD(registerForRemoteNotifications)
@end

@interface RCT_EXTERN_REMAP_MODULE(LeanplumInbox, RNLeanplumInbox, RCTEventEmitter)
RCT_EXTERN_METHOD(getInbox: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getMessage: (NSString *)messageId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(read: (NSString *)messageId)
RCT_EXTERN_METHOD(markAsRead: (NSString *)messageId)
RCT_EXTERN_METHOD(remove: (NSString *)messageId)
RCT_EXTERN_METHOD(onChanged: (NSString *)listener)
RCT_EXTERN_METHOD(onForceContentUpdate: (NSString *)listener)
@end

File: ios/Leanplum.swift
//
//  Leanplum.swift
//  Leanplum
//
//  Created by Alik . Risco on 30.01.20.
//  Copyright © 2023 Leanplum. All rights reserved.
//

import Foundation
import Leanplum
import CleverTapSDK

@objc(RNLeanplum)
class RNLeanplum: RCTEventEmitter {
    
    enum Constants {
        static let DatePrefix = "lp_date_"
        static let SupportedEvents = [
            "ctReadyEvent"
        ]
    }

    var variables: [String: LeanplumVar] = [:]
    var allSupportedEvents: [String] = Constants.SupportedEvents
    
    @objc
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override func supportedEvents() -> [String] {
        return self.allSupportedEvents
    }
    
    @objc
    func setAppVersion(_ appVersion: String) {
        Leanplum.setAppVersion(appVersion)
    }
    
    @objc
    func setAppIdForDevelopmentMode(_ appId: String, accessKey: String) {
        Leanplum.setAppId(appId, developmentKey: accessKey)
    }
    
    
    @objc
    func setAppIdForProductionMode(_ appId: String, accessKey: String) {
        Leanplum.setAppId(appId, productionKey: accessKey)
    }

    @objc
    func setApiConnectionSettings(_ hostName: String, apiPath: String, ssl: Bool) {
        Leanplum.setApiHostName(hostName, apiPath: apiPath, ssl: ssl)
    }

    @objc
    func setSocketConnectionSettings(_ hostName: String, port: Int32) {
        Leanplum.setSocketHostName(hostName, port: port)
    }
    
    @objc
    func setDeviceId(_ id: String) {
        Leanplum.setDeviceId(id)
    }
    
    @objc
    func setUserId(_ id: String) {
        Leanplum.setUserId(id)
    }
    
    @objc
    func setUserAttributes(_ attributes: NSDictionary) {
        guard let attributesDict = attributes as? Dictionary<String, Any> else {
            return
        }

        let attr = attributesDict.mapValues {
            if let str = $0 as? String, str.hasPrefix(Constants.DatePrefix) {
                let index = str.index(str.startIndex, offsetBy: Constants.DatePrefix.count)
                let ts = String(str[index...])
                if let ti = Double(ts) {
                    return Date(timeIntervalSince1970: ti/1000) as Any
                }
            }
            return $0
        }

        DispatchQueue.main.async {
            Leanplum.setUserAttributes(attr)
        }
    }

    @objc
    func userId(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(Leanplum.userId())
    }

    @objc
    func deviceId(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(Leanplum.deviceId())
    }

    @objc
    func start() {
        Leanplum.start()
    }

    @objc
    func hasStarted(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(Leanplum.hasStarted())
    }
    
    @objc
    func track(_ event: String, params: NSDictionary) {
        guard let parametersDict = params as? Dictionary<String, Any> else {
            return
        }
        Leanplum.track(event, params: parametersDict)
    }
    
    @objc
    func trackPurchase(_ purchaseEvent: String, value: Double, currencyCode: String, purchaseParams: NSDictionary) {
        guard let parametersDict = purchaseParams as? Dictionary<String, Any> else {
            return
        }
        Leanplum.track(event: purchaseEvent, value: value, currencyCode: currencyCode, params: parametersDict)
    }
    
    @objc
    func trackInAppPurchases() {
        Leanplum.trackInAppPurchases()
    }
    
    
    @objc
    func disableLocationCollection() {
        Leanplum.disableLocationCollection()
    }
    
    @objc
    func setDeviceLocation(_ latitude: Double, longitude: Double, type: UInt) {
        if let accuracyType = Leanplum.LocationAccuracyType(rawValue: type) {
            Leanplum.setDeviceLocation(latitude: latitude, longitude: longitude, type: accuracyType)
        }
    }
    
    @objc
    func forceContentUpdate() {
        Leanplum.forceContentUpdate()
    }
    
    
    @objc
    func setVariables(_ variables: NSDictionary) {
        guard let variablesDict = variables as? Dictionary<String, Any> else {
            return
        }
        for (key, value) in variablesDict {
            if let lpVar = LeanplumTypeUtils.createVar(key: key, value: value) {
                self.variables[key] = lpVar
            }
        }
    }
    
    @objc
    func getVariable(_ variableName: String, resolver resolve: RCTPromiseResolveBlock,
                     rejecter reject: RCTPromiseRejectBlock
    ) {
        if let lpVar = self.variables[variableName] {
            resolve(lpVar.value)
        } else {
            resolve(nil)
        }
    }
    
    @objc
    func getVariables(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(self.getVariablesValues())
    }
    
    func getVariablesValues() -> [String: Any] {
        var allVariables = [String: Any]()
        for (key, value) in self.variables {
            if(value.kind == "file") {
                continue
            }
            allVariables[key] = value.value
        }
        return allVariables
    }
    
    @objc
    func onStartResponse(_ callback: @escaping RCTResponseSenderBlock) {
        Leanplum.onStartResponse { (success:Bool) in
            callback([success])
        }
    }
    
    @objc
    func onValueChanged(_ variableName: String) {
        if let lpVar = self.variables[variableName] {
            self.allSupportedEvents.append(variableName)
            lpVar.onValueChanged { [weak self] in
                self?.sendEvent(withName: variableName, body: lpVar.value)
            }
        }
    }
    
    @objc
    func onVariablesChanged(_ listener: String) {
        self.allSupportedEvents.append(listener)
        Leanplum.onVariablesChanged { [weak self] in
            self?.sendEvent(withName: listener, body: self?.getVariablesValues())
        }
    }
    
    @objc
    func setVariableAsset(_ name: String, filename: String) {
        self.allSupportedEvents.append(name)
        let lpVar = Var(name: name, file: filename)
        self.variables[name] = lpVar
        lpVar.onFileReady({ [weak self] in
            self?.sendEvent(withName: name, body: lpVar.fileValue())
        })
    }
    
    @objc
    func getVariableAsset(_ name: String,
                          resolver resolve: RCTPromiseResolveBlock,
                          rejecter reject: RCTPromiseRejectBlock) {
        if let lpVar = self.variables[name] {
            resolve(lpVar.fileValue())
        } else {
            resolve(nil)
        }
    }
    
    @objc
    func pauseState() {
        Leanplum.pauseState()
    }
    
    @objc
    func resumeState() {
        Leanplum.resumeState()
    }
    
    @objc
    func advanceTo(_ state: String) {
        DispatchQueue.main.async {
            Leanplum.advance(state: state)
        }
    }
    
    @objc
    func advanceToWithInfo(_ state: String, info: String) {
        DispatchQueue.main.async {
            Leanplum.advance(state: state, info: info)
        }
    }
    
    @objc
    func advanceToWithParams(_ state: String, params: NSDictionary) {
        guard let paramsDict = params as? Dictionary<String, Any> else {
            return
        }
        DispatchQueue.main.async {
            Leanplum.advance(state: state, params: paramsDict)
        }
    }
    
    
    @objc
    func advanceToWithInfoAndParams(_ state: String, info: String, params: NSDictionary) {
        guard let paramsDict = params as? Dictionary<String, Any> else {
            return
        }
        DispatchQueue.main.async {
            Leanplum.advance(state: state, info: info, params: paramsDict)
        }
    }

    @objc
    func onVariablesChangedAndNoDownloadsPending(_ listener: String) {
        self.allSupportedEvents.append(listener)
        Leanplum.onVariablesChangedAndNoDownloadsPending { [weak self] in
            self?.sendEvent(withName: listener, body: nil)
        }
    }

    @objc
    func onceVariablesChangedAndNoDownloadsPending(_ listener: String) {
        self.allSupportedEvents.append(listener)
        Leanplum.onceVariablesChangedAndNoDownloadsPending { [weak self] in
            self?.sendEvent(withName: listener, body: nil)
        }
    }

    @objc
    func onMessageDisplayed(_ listener: String?) {
        if let listener = listener {
            self.allSupportedEvents.append(listener)
            ActionManager.shared.onMessageDisplayed { [weak self] actionContext in
                self?.sendEvent(withName: listener, body: actionContext.dictionary)
            }
        } else {
            ActionManager.shared.onMessageDisplayed(nil)
        }
    }

    @objc
    func onMessageDismissed(_ listener: String?) {
        if let listener = listener {
            self.allSupportedEvents.append(listener)
            ActionManager.shared.onMessageDismissed { [weak self] actionContext in
                self?.sendEvent(withName: listener, body: actionContext.dictionary)
            }
        } else {
            ActionManager.shared.onMessageDismissed(nil)
        }
    }

    @objc
    func onMessageAction(_ listener: String?) {
        if let listener = listener {
            self.allSupportedEvents.append(listener)
            ActionManager.shared.onMessageAction { [weak self] actionName, context in
                self?.sendEvent(withName: listener, body: context.dictionary)
            }
        } else {
            ActionManager.shared.onMessageAction(nil)
        }
    }

    @objc func registerForRemoteNotifications() {
        Leanplum.enablePushNotifications()
    }
    
    @objc func securedVars(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        guard let securedVars = Leanplum.securedVars() else {
            resolve(nil)
            return
        }

        var securedVarsDictionary: [String: Any] = [:]
        securedVarsDictionary["json"] = securedVars.varsJson()
        securedVarsDictionary["signature"] = securedVars.varsSignature()
        resolve(securedVarsDictionary)
    }

    @objc
    func isQueuePaused(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(ActionManager.shared.isPaused)
    }

    @objc
    func setQueuePaused(_ paused: Bool) {
        ActionManager.shared.isPaused = paused
    }

    @objc
    func isQueueEnabled(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve(ActionManager.shared.isEnabled)
    }

    @objc
    func setQueueEnabled(_ enabled: Bool) {
        ActionManager.shared.isEnabled = enabled
    }
    
    @objc
    func onCleverTapInstance(_ listener: String?) {
        if let listener = listener {
            self.allSupportedEvents.append(listener)
            Leanplum.addCleverTapInstance(callback: CleverTapInstanceCallback(callback: { [weak self] instance in
                self?.sendEvent(withName: listener, body: instance.config.accountId)
            }))
        }
    }

    @objc
    func migrationConfig(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        var migrationConfig:[String: Any] = [:]
        migrationConfig["state"] = MigrationManager.shared.state.description
        migrationConfig["accountId"] = MigrationManager.shared.cleverTapAccountId
        migrationConfig["accountToken"] = MigrationManager.shared.cleverTapAccountToken
        migrationConfig["accountRegion"] = MigrationManager.shared.cleverTapAccountRegion
        migrationConfig["attributeMappings"] = MigrationManager.shared.cleverTapAttributeMappings
        migrationConfig["identityKeys"] = MigrationManager.shared.cleverTapIdentityKeys
        resolve(migrationConfig)
    }
}
